### PR TITLE
8331264: Reduce java.lang.constant initialization overhead

### PR DESCRIPTION
**Pull request by cl4es**
I'm looking at ways at reducing/eliminating startup overheads the classfile API in preparation of #17108, and have pushed a series of enhancements to that effect already. This PR is a collection of minor improvements which add up to a 1.5% reduction in retired instructions - or a 5% reduction in executed bytecode - on a simple lambda startup test.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331264](https://bugs.openjdk.org/browse/JDK-8331264): Reduce java.lang.constant initialization overhead (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18991/head:pull/18991` \
`$ git checkout pull/18991`

Update a local copy of the PR: \
`$ git checkout pull/18991` \
`$ git pull https://git.openjdk.org/jdk.git pull/18991/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18991`

View PR using the GUI difftool: \
`$ git pr show -t 18991`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18991.diff">https://git.openjdk.org/jdk/pull/18991.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18991#issuecomment-2082126830)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/constant/MethodTypeDescImpl.java [hunk 1]
```diff
@@ -66,12 +66,12 @@ private MethodTypeDescImpl(ClassDesc returnType, ClassDesc[] validatedArgTypes)
...
      */
     static MethodTypeDescImpl ofTrusted(ClassDesc returnType, ClassDesc[] trustedArgTypes) {
-        Objects.requireNonNull(returnType);
+        requireNonNull(returnType);
         if (trustedArgTypes.length == 0) // implicit null check
             return new MethodTypeDescImpl(returnType, ConstantUtils.EMPTY_CLASSDESC);
 
         for (ClassDesc cd : trustedArgTypes)
-            if (cd.isPrimitive() && cd.descriptorString().charAt(0) == 'V') // implicit null check
+            if (cd.descriptorString().equals("V")) // implicit null check
```
Comment (liach) [inline src/java.base/share/classes/java/lang/constant/MethodTypeDescImpl.java:74]: Just curious, is `equals` faster than `charAt(0) ==` for single-char strings?

**Review by liach — APPROVED** [review]

File: src/java.base/share/classes/java/lang/constant/ConstantDescs.java [hunk 2]
```diff
@@ -64,125 +64,125 @@ private ConstantDescs() { }
...
 
     /** {@link ClassDesc} representing {@link ConstantBootstraps} */
-    public static final ClassDesc CD_ConstantBootstraps = ClassDesc.of("java.lang.invoke.ConstantBootstraps");
+    public static final ClassDesc CD_ConstantBootstraps = new ReferenceClassDescImpl("Ljava/lang/invoke/ConstantBootstraps;");
 
     private static final ClassDesc[] INDY_BOOTSTRAP_ARGS = {
-            ConstantDescs.CD_MethodHandles_Lookup,
-            ConstantDescs.CD_String,
-            ConstantDescs.CD_MethodType};
+            CD_MethodHandles_Lookup,
```
Comment (minborg) [inline src/java.base/share/classes/java/lang/constant/ConstantDescs.java:178]: Can these fields be @Stable?

Comment (cl4es) [reply src/java.base/share/classes/java/lang/constant/ConstantDescs.java:178]: I don't think it'd hurt, but I would be surprised if any of these would be hot enough to see a benefit from that.

Comment (liach) [reply src/java.base/share/classes/java/lang/constant/ConstantDescs.java:178]: I don't think this really helps; the array elements are not used directly via field reference.

**Review by mlchung — COMMENTED** [review]
The changes look good to me but I wonder if the non-zero length check before calling `arraycopy` really needed?  That seems to add some noise to the code.

**Comment by cl4es** [comment]
> The changes look good to me but I wonder if the non-zero length check before calling `arraycopy` really needed? That seems to add some noise to the code.

I recall benchmarking similar code in `MethodType` extensively, and found that it had a small but significant impact, especially when adding or dropping parameters at the end or beginning of an argument list (which is the most common case). Adhoc JMH:

```
    @Benchmark
    public MethodTypeDesc dropFirstParameter() {
        return desc.dropParameterTypes(0, 1);
    }

Name                                                                   (descString) Cnt   Base   Error    Test   Error  Unit  Change
MethodTypeDescFactories.dropFirstParameter  (Ljava/lang/Object;Ljava/lang/String;)I   6 20,382 ± 0,635  16,116 ± 0,737 ns/op   1,26x (p = 0,000*)
MethodTypeDescFactories.dropFirstParameter ([IJLjava/lang/String;Z)Ljava/util/List;   6 26,875 ± 0,652  19,863 ± 0,749 ns/op   1,35x (p = 0,000*)
MethodTypeDescFactories.dropFirstParameter                                 (..IIJ)V   6 29,774 ± 0,983  20,747 ± 1,181 ns/op   1,44x (p = 0,000*)
MethodTypeDescFactories.dropFirstParameter                 (.....................).   6 33,224 ± 0,939  30,350 ± 1,739 ns/op   1,09x (p = 0,000*)
  * = significant
```

**Review by mlchung — APPROVED** [review]

**Pull request closed without merging**